delivery_roulier/models/stock_picking.py

```diff
@@ -0,0 +1,282 @@
+#  @author Raphael Reverdy <raphael.reverdy@akretion.com>
+#          David BEAL <david.beal@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from datetime import date, timedelta
+
+from odoo import fields, models
+from odoo.exceptions import UserError
+from odoo.tools.translate import _
+
+from ..decorator import implemented_by_carrier
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    # API:
+
+    @implemented_by_carrier
+    def _get_sender(self, package=None):
+        pass
+
+    @implemented_by_carrier
+    def _get_receiver(self, package=None):
+        pass
+
+    @implemented_by_carrier
+    def _get_shipping_date(self, package):
+        pass
+
+    @implemented_by_carrier
+    def _get_account(self, package=None):
+        pass
+
+    @implemented_by_carrier
+    def _get_auth(self, account, package=None):
+        pass
+
+    @implemented_by_carrier
+    def _get_service(self, account, package=None):
+        pass
+
+    @implemented_by_carrier
+    def _convert_address(self, partner):
+        pass
+
+    @implemented_by_carrier
+    def _get_label_format(self, account):
+        pass
+
+    @implemented_by_carrier
+    def _get_from_address(self):
+        pass
+
+    @implemented_by_carrier
+    def _get_to_address(self):
+        pass
+
+    @implemented_by_carrier
+    def _cancel_shipment(self):
+        pass
+
+    # End of API.
+
+    # Implementations for base_delivery_carrier_label
+    def _is_roulier(self):
+        self.ensure_one()
+        return self.carrier_id._is_roulier()
+
+    def _roulier_generate_labels(self):
+        """
+        Return format expected by send_shipping : a list of dict (one dict per
+        picking).
+        {
+            'exact_price': 0.0,
+            'tracking_number': "concatenated numbers",
+            'labels': list of dict of labels, managed by base_delivery_carrier_label
+        }
+        """
+        label_info = []
+        for picking in self:
+            move_line_no_pack = picking.move_line_ids.filtered(
+                lambda ml: ml.qty_done > 0.0 and not ml.result_package_id
+            )
+            if move_line_no_pack:
+                # For automatic package creation, for use cases where we only have one
+                # pack per picking and we don't want to force the user to use pack
+                # consider installing delivery_automatic_package OCA module
+                raise UserError(
+                    _(
+                        "Some products have no destination package in picking %s, "
+                        "please add a destination package in order to be able to "
+                        "generate the carrier label."
```

Comment:
You might suggest to install `delivery_automatic_package` in the raise ?
Function users won't read source code.

Reply:
This is not really an information for a functional/final user, I don't thunk it should be in the odoo error. Also because it can be consitent for some use cases and for other, they just want it to raise when user forgot to put a pack and it makes no sense to always tell him to install a new module!

Anyway, I am aware it is not very visible but that was just to put the information somewhere. Maybe I could put it in the know issue part instead

Reply:
After thinking of it, wouldn't it be more correct to add a new module like delivery_carrier_package_required ?

Reply:
It is really 3 lines of code, I am not sure it is worth a module ?
Why would that help ? (I am not specially against, I am just wondering the gain ?)

Reply:
I would say it sounds like a transversal requirement for several modules.

BUT, as the implementations can vary in each of them, it could become too harsh to raise the error at the good moment.

So, let's the error implementation in the modules that require it.

Reply:
@bealdav I have removed the comment and added the information in the README.